add Yaw/Heading to MAV_CMD_SET_GLOBAL_ORIGIN

njdevi11 opened the pull request:
add Yaw/Heading to MAV_CMD_SET_GLOBAL_ORIGIN to indicate which direction the vehicle is facing on initialization if compass is unavailable.

hamishwillee commented on message_definitions/v1.0/development.xml:
```diff
@@ -194,7 +194,7 @@
         <param index="2">Empty</param>
         <param index="3">Empty</param>
-        <param index="4">Empty</param>
+        <param index="4" label="Yaw" units="rad">Yaw heading. NaN to indicate Invalid/Unknown.</param>
```
```suggestion
        <param index="4" label="Yaw" units="rad">Yaw heading (relative to North). NaN to indicate Invalid/Unknown.</param>
```

hamishwillee commented on message_definitions/v1.0/development.xml:
```diff
@@ -194,7 +194,7 @@
         <param index="2">Empty</param>
         <param index="3">Empty</param>
-        <param index="4">Empty</param>
+        <param index="4" label="Yaw" units="rad">Yaw heading (relative to North). NaN to indicate Invalid/Unknown.</param>
```
```suggestion
        <param index="4" label="Yaw" units="rad">Yaw heading (relative to North). NaN to indicate invalid/unknown.</param>
```

hamishwillee commented on message_definitions/v1.0/development.xml:
```diff
@@ -194,7 +194,7 @@
         <param index="2">Empty</param>
         <param index="3">Empty</param>
-        <param index="4">Empty</param>
+        <param index="4" label="Yaw" units="rad">Yaw heading (relative to North). NaN to indicate invalid/unknown.</param>
```
Many systems will be setting this to zero by default, while some will set it as NaN. What this means is that this is likely to be set to 0 by lots of systems, and they will mean "I didn't know this field had been updated" not "Vehicle is facing north". 

@peterbarker @auturgy @julianoes What do you think about this? Is it useful to be able to set the yaw as part of the global origin, and presumably thereby align the local frame to the global frame?

bresch replied:
The position of the origin is not the position of the vehicle. If you add a heading field in here, it would be to have the local frame different from NED; which is not what you want I think.
If I understood correctly, you want to set the heading of the vehicle manually, right? In this case you should rather add a field to https://mavlink.io/en/messages/common.html#MAV_CMD_EXTERNAL_POSITION_ESTIMATE which is currently used to give manual position updates to the estimator.

njdevi11 replied:
@bresch Excellent point. The goal here is to set the initial position and orientation of where the vehicle will start or has started to align the local frame to NED, rather than assume we already know NED. The application would be complete gps/magnetometer less operation such as SLAM mapping.

I agree that MAV_CMD_EXTERNAL_POSITION_ESTIMATE would be a good place to add a heading field to express the current orientation of the vehicle while in operation. MAV_CMD_EXTERNAL_POSITION_ESTIMATE would be the way to correct drift from a gyro, but that is still separate from expressing the orientation of the vehicle when it started. We should also add param4 as heading to be consist across position messages like MAV_CMD_DO_REPOSITION even if it's not used on some systems.

@hamishwillee You can not prevent the user entering incorrect information as long as it passes a sanity check... e.g. -2pi to 2pi, garbage in, garbage out. The best we can do is allow the user to correct a mistake. The worse case scenario is that the local position / coordinates rotate around the initial position.

peterbarker commented:
The vehicle may never actually be at its navigation origin, so I don't think an orientation in the message makes sense.

Telling the vehicle definitively via a `MAV_CMD_EXTERNAL_POSITION_ESTIMATE` might be an option - but we actually already have a lot of ways to inject that information already with the various *_POSITION_ESTIMATE messages.

njdevi11 commented:
@peterbarker  How do you suggest we orient the vehicle to the global frame on initialization? Should we send a MAV_CMD_DO_SET_GLOBAL_ORIGIN, and then MAV_CMD_EXTERNAL_POSITION_ESTIMATE with the heading? I see your point that rotating the local coordinate frame could be complicated. But we should be able to do what we need as long as heading is added to MAV_CMD_EXTERNAL_POSITION_ESTIMATE.

peterbarker commented:
> @peterbarker How do you suggest we orient the vehicle to the global frame on initialization? Should we send a MAV_CMD_DO_SET_GLOBAL_ORIGIN, and then MAV_CMD_EXTERNAL_POSITION_ESTIMATE with the heading? I see your point that rotating the local coordinate frame could be complicated. But we should be able to do what we need as long as heading is added to MAV_CMD_EXTERNAL_POSITION_ESTIMATE.

I don't think we can actually use "position_estimate"; if we want to inject attitude information then we should inject attitude information, not just roll/pitch.

I would suggest a new command, `MAV_CMD_ATTITUDE_ESTIMATE`.  While autopilots will have code paths for existing messages like GLOBAL_POSITION_ESTIMATE, VICON_POSITION_ESTIMATE and VISION_POSITION_ESTIMATE, those are usually used in a continuous-input scenario.  I believe this new command would be more of a reset, this is your attitude" message.

The fact that at least the yaw estimate will probably instantly disagree with magnetometer sensor data - that could be an issue...

Again, you don't need to specify a navigation origin before telling the vehicle what its roll/pitch and yaw are - a position doesn't help provide a frame for an attitude and I don't think any vehicle flies with an attitude frame which isn't earth!  (leaving aside some weirdness for tailsitters)

I think all you need to do is throw the command at the vehicle and it can reset to that attitude as it sees fit.

njdevi11 commented:
@peterbarker I think you're mistaking the desire here, I've never once mentioned setting roll and pitch and don't need to inject roll/pitch. We need to orient a vehicle that does not have or is not using it's magnetometer so it knows what direction it's facing in the earth frame. It will never conflict with the magnetometer because there is none. This will be done by either selecting two points on a map, or casting a line on a map based on an input from the user.

I am currently working on at least two systems that do not use a magnetometer for navigation that are not PX4 or APM based. The action of providing a yaw to the autopilot will allow the autopilot to correctly orient it's NED and global position messages. Or if the heading is incorrect after mapping has started, the operator of the vehicle needs to align the created map and vehicle orientation with google/osm maps.

Jaeyoung-Lim commented:
Your global frame by definition, is aligned with NED, and therefore the frame origin should not depend on either the heading estimate of the vehicle, nor whether the existence of a magnetometer.

I think what you really want is someway to initialize or set the vehicle yaw estimate.

peterbarker commented:
On Sat, 7 Jun 2025, njdevi11 wrote:

> @peterbarker I think you're mistaking the desire here, I've never once mentioned setting roll and pitch and don't need to inject roll/pitch. We need to orient a vehicle that does not

No, I think you're describing exactly what I was thinking you wanted.

Pass through the pitch/roll you're not using as NaN ("don't know").  Or 
if you're catapult launching, use the rail angle as pitch, I guess.

> have or is not using it's magnetometer so it knows what direction it's facing in the earth frame. It will never conflict with the magnetometer because there is none. This will be done by
> either selecting two points on a map, or casting a line on a map based on an input from the user.

I was imagining either no mag or a mag compromised by its takeoff 
surrounds.

> I am currently working on at least two systems that do not use a magnetometer for navigation that are not PX4 or APM based. The action of providing a yaw to the autopilot will allow the
> autopilot to correctly orient it's NED and global position messages. Or if the heading is incorrect after mapping has started, the operator of the vehicle needs to align the created map
> and vehicle orientation with google/osm maps.

Right.  Externally setting the *position* of the vehicle is done with the 
message right next to where you I am suggesting you put this new command. 
Similarly with external wind estimate.

Your scenario here is interesting.  We have many users flying without 
compasses.  While having this message might help them, a lot of them are 
flying sans any sorts of tools which might let them supply the heading 
they're just about to throw the aircraft off at.  Which is a shame, 
because the whole, "think quick" to get an initial heading (usually done 
when a suitable GPS velocity has been obtained isn't 100% reliable.

I should also mention that there's a related (but unsuitable) message used 
by ArduPilot to aid with calibrating magnetomers on large or unwieldy 
vehicles; note I'm saying "calibration" here, not the sort of thing you 
use at runtime.  Just noting it.

njdevi11 commented:
@Jaeyoung-Lim Yes, this is correct, Im not suggesting redefining the NED reference to something different, I'm suggesting having the ability to correct it when it's not correct. for example, I start the vehicle with no gps and no mag, facing 90*, the vehicle assumes it's at 0,0 and facing north. In the simplest case, before we arm and start moving, we send a message, or set of messages to let the AP know that we are in a particular location, lat/lon/alt and facing this direction. So now the AP will know if I move forward 100m I have moved 100m East, instead of North. In a more complicated example, I have moved 100m east, and the ap reports that I'm at 0,0 + 100m north. I notice the problem and now need to send a message that I really started at a location and heading. The expected result is that the position estimate will move and rotate. both scenarios result in the same end state.

@peterbarker I'm not understanding what your solution/proposal is. I imagine it would be a separate message(s) that includes Lat/Lon/Alt/Heading and can either be used to describe where the vehicle currently is, or it started based on the scenario I gave above. I don't see the need to add new messages since they are mostly redundant with the MAV_CMD_DO_SET_GLOBAL_ORIGIN, and MAV_CMD_EXTERNAL_POSITION_ESTIMATE if they included an option heading.

Based on your recommendation and if what I read back is correct I can add a new message, but I think this will add unneeded ambiguity.

Jaeyoung-Lim commented:
> @Jaeyoung-Lim Yes, this is correct, Im not suggesting redefining the NED reference to something different, I'm suggesting having the ability to correct it when it's not correct. for example, I start the vehicle with no gps and no mag, facing 90*, the vehicle assumes it's at 0,0 and facing north. In the simplest case, before we arm and start moving, we send a message, or set of messages to let the AP know that we are in a particular location, lat/lon/alt and facing this direction. 

There is nothing to correct, because the global frame is always aligned with north and east. The problem is that you are assuming that the vehicle is facing north when you don't know the vehicle heading. Therefore, you should not try to move the global or local frame, you should try to adjust the vehicle heading estimate based on external reference.

peterbarker commented:
FWIW, quick discussion with tridge and he noted that we would probably treat the input from this message in a qay very similar to the way we do a yaw estimate coming from a moving baseline setup.

He did ask whether covariances - which was a good point.

njdevi11 commented:
> > @Jaeyoung-Lim Yes, this is correct, Im not suggesting redefining the NED reference to something different, I'm suggesting having the ability to correct it when it's not correct. for example, I start the vehicle with no gps and no mag, facing 90*, the vehicle assumes it's at 0,0 and facing north. In the simplest case, before we arm and start moving, we send a message, or set of messages to let the AP know that we are in a particular location, lat/lon/alt and facing this direction.
> 
> There is nothing to correct, because the global frame is always aligned with north and east. The problem is that you are assuming that the vehicle is facing north when you don't know the vehicle heading. Therefore, you should not try to move the global or local frame, you should try to adjust the vehicle heading estimate based on external reference.

You're correct, But I wasn't suggesting to change the global frame. I Looked a bit deeper and found this message. LOCAL_POSITION_NED_SYSTEM_GLOBAL_OFFSET: "The offset in X, Y, Z and yaw between the LOCAL_POSITION_NED messages" This yaw offset is what we need to adjust. Is there an existing message that allows us to do that? Adjusting this based on the current position of the vehicle seems more difficult and less intuitive than just adjusting it at it's source.

@peterbarker I'll need some time to think about how MAV_CMD_EXTERNAL_ATTITUDE_ESTIMATE will be handled and used to update the global coordinates and history of the vehicle. If you assume there is no drift in the yaw then this is reasonable to accomplish what I mentioned above in this reply, but I think it's still necessary to be able to adjust the initial yaw offset and update the current attitude/yaw to correct drift.

njdevi11 commented:
@peterbarker @Jaeyoung-Lim 
How do we correct the yaw offset reported in the LOCAL_POSITION_NED_SYSTEM_GLOBAL_OFFSET message? This is the offset needed to adjust and the original intent of this MR.

peterbarker commented:
On Wed, 25 Jun 2025, njdevi11 wrote:

> @peterbarker @Jaeyoung-Lim
> How do we correct/update the yaw offset reported in the LOCAL_POSITION_NED_SYSTEM_GLOBAL_OFFSET message? This is the offset needed to adjust and the original intent of this MR.

I'm not familiar with that message.  I didn't even realise you could have 
your local co-ordinate system at an arbitrary angle from the global 
coordinate system!  In fact, I do not understand that message - I'm 
wondering if they're missing a reference to the global origin in there 
somewhere.

Pretty sure this is *not* what you are wanting, anyway.

You are not wanting to rotate a local coordinate system, you are trying to 
tell the vehicle its yaw in the global coordinate system.  (to be clear, a 
robot can navigate in several different coordinate systems.  If you have 
a GPS you'll be navigating in the global frame for the most part.  If you 
don't have a GPS you will navigate in a local frame.

njdevi11 commented:
We do want to rotate the local coordinate system in the global reference or have the ability to update the current position based on a similar concept if we do not actually rotate the local coordinate system. Something like rotate the local frame, determine where the vehicle would be in global frame, then invert that command, and move the vehicle's local coords and heading to where the new position of the vehicle is based on the original reference. The navigation system will be able to take in coordinates via lat/lon or local coords. e.g. once the global orientation is properly set. we can tell the vehicle to go to this lat/lon, or we can tell the vehicle to move local x=x+10m. Under the hood this usually implemented as the client telling the system to navigate to a certain lat/lon and then this is converted to a local position on the vehicle. The vehicle is expected to report it's correct position in Lat/Lon via GLOBAL_POSITION_INT when it has enough information to do so.

Your point about updating the yaw of the vehicle is valid, we might still need to update/correct/nudge the yaw of the vehicle due to "wheel slip" or gyro/sensor drift. But it's more pressing that we are able to take a map that was created without any reference and correctly orient it in the global space.

I wasn't aware of LOCAL_POSITION_NED_SYSTEM_GLOBAL_OFFSET either until I dug in deeper but it seems very applicable to what we are doing. I'm reluctant to add more messages because there are too many redundant and ambiguous messages and this just leads to mis interpretation of the intent of those messages as the descriptions are rarely iron clad.

I'm a bit confused about the LOCAL_POSITION_NED message. Is this always North East Down in relation to the global reference as it seems like you are suggesting? Or is this North East Down (as opposed to Cartesian format which would be East North Up?) in relation to the Local Frame I believe typically this frame is aligned with north, but LOCAL_POSITION_NED_SYSTEM_GLOBAL_OFFSET suggests that it should be interpreted that LOCAL_POSITION_NED is in relation to the frame of the vehicle and not the globe. I think we need to define whether or not it is possible/acceptable to rotate the local frame in the global frame, and if not why does LOCAL_POSITION_NED_SYSTEM_GLOBAL_OFFSET exist?

Perhaps we can discuss on the next mavlink developer call? I can invite some colleagues to discuss as well.

josh-raanan-skydio commented:
Having a heading included in MAV_CMD_EXTERNAL_POSITION_ESTIMATE would be a great addition. This is exactly what Skydio is doing for our [Anchor Point](https://support.skydio.com/hc/en-us/articles/36550953430811-How-to-use-Anchor-Point-on-Skydio-X10D) feature. Having a command with lat,lon,yaw would enable us to easily extend Mavlink support to the feature.

peterbarker commented:
On Fri, 27 Jun 2025, josh-raanan-skydio wrote:

> Having a heading included in MAV_CMD_EXTERNAL_POSITION_ESTIMATE would be a
> great addition. This is exactly what Skydio is doing for our Anchor Point
> feature. Having a command with lat,lon,yaw would enable us to easily extend
> Mavlink support to the feature.

And your thoughts on MAV_CMD_EXTERNAL_ATTITUDE_ESTIMATE?

hamishwillee commented:
@peterbarker Response to https://github.com/mavlink/mavlink/pull/2282#issuecomment-3045534050

My leaning is to do MAV_CMD_EXTERNAL_ATTITUDE_ESTIMATE unless there really is no use case "ever" for passing the pitch / roll estimate.
